### PR TITLE
Fixing parameter decorators

### PR DESCRIPTION
**Pull request by doktordirk**
The previous fix for autoinject and inheritance compared constructor parameter types with those of the inject property. That broke the  parameter decorators.

This PR reverts autoinject back to the previous assigning of the parameter decorators to static inject by index and just ensures that only the own property 'inject' is used for autoinject and parameter decorators to avoid the inheritance issues. Since the parameter decorators (including inject as parameter decorator) also act as autoinject and manipulate static inject, this was simpified and clarified by letting them use autoinject and rename appropriately.

Included are the fixed test for parameter decorators (including inject) for the use with autoinject, tests for inheritance together with parameter decorators, and a test for the checkparent option of @optional

### CODE REVIEW

**Comment by zewa666** [comment]
This is awesome, @Eisenbergeffect and @bigopon can you guys give it a look as well?

File: src/injection.js [hunk 1]
```diff
@@ -6,24 +6,16 @@ import {_emptyParameters} from './container';
...
-      for (let i = 0; i < autoInject.length; i++) {
-        //check if previously injected.
-        if (previousInject[i] && previousInject[i] !== autoInject[i]) {
-          const prevIndex = previousInject.indexOf(autoInject[i]);
-          if (prevIndex > -1) {
-            previousInject.splice(prevIndex, 1);
-          }
-          previousInject.splice((prevIndex > -1 && prevIndex < i) ? i - 1 : i, 0, autoInject[i]);
-        } else if (!previousInject[i]) {//else add
+      for (let i = 0; i++; i < autoInject.length) {
```
Comment (EisenbergEffect) [inline src/injection.js]: This `for` loop doesn't look correct...

File: src/resolvers.js [hunk 2]
```diff
@@ -315,6 +315,9 @@ export class NewInstance {
 }
 
 export function getDecoratorDependencies(target, name) {
+  if (!target.hasOwnProperty('inject')) {
```
Comment (EisenbergEffect) [inline src/resolvers.js]: Is the idea here to effectively block the inheritance chain?

File: src/injection.js (hunk 1 above)
Comment (doktordirk) [reply src/injection.js]: argh! though highlights that this part isn't need at all. reason is that the parameter decorators act as autoinject themself. hence that faulty loop wasn't picked up in the original implementation.  (was picked up and fixed in following commits, so shame on me for trying to reintroduce it).

so, with the currently tested scenarions, following would be enough
```js
export function autoinject(potentialTarget?: any): any {
  let deco = function(target) {
    if (!target.hasOwnProperty('inject')) {
      target.inject = metadata.getOwn(metadata.paramTypes, target) || _emptyParameters;
    }
// else a parameter decorator has done it already
  };

  return potentialTarget ? deco(potentialTarget) : deco;
}
```
and atm, i can't think of any scenario where more would be needed.

File: src/resolvers.js (hunk 2 above)
Comment (doktordirk) [reply src/resolvers.js]: yes, as the parameter decorators act as autoinject themself (adding @autoinject would just be for clarity) it needs the same autoinject. and with autoinject, one never wants to use the inherited one. after all, autoinject, takes it all from the current constructor parameters.
i'll rewrite it, to make it clearer, that it is exactly the same

**Comment by doktordirk** [comment]
@EisenbergEffect see anwers in the code review. 
also, i removed the test for an existing own static inject function as neither function nor array make sense. however, either would cause trouble. only way i see to test/prevent it, to add metadata to inject indicating whether the user or autoinject/parameter decorators have added that property. i probably will add that for consideration

**Comment by doktordirk** [comment]
also i'd like to rename getDecoratorDependencies to getOrCreateStaticInject. though as getDecoratorDependencies got exported that would be a breaking change in theory, even though it's rather unlikely anyone used in his own code

**Comment by doktordirk** [comment]
I added a check against accidentally using autoinject together with static inject to prevent following

```js
      @autoinject
      class App {
        static inject: [Container];
        constructor(logger: Logger) {
          this.logger = logger;
        }
      } 
```

It would be unclear whether the user expect logger to be a Container or a Logger. Current implementation would make a Container. Better just to avoid that, though I don't like setting metadata on everything I again didn't see much of another solution.

**Comment by doktordirk** [comment]
as for the last, there would be an error naturally, once one try to use this.logger with something that doesn't exist on that type. hence, preventing that scenario early with a test might not be that important. i'll revert it for now to have cleaner code

**Comment by doktordirk** [comment]
@EisenbergEffect got a lot shorter 😄

File: src/resolvers.js [hunk 3]
```diff
@@ -315,25 +316,18 @@ export class NewInstance {
 }
 
 export function getDecoratorDependencies(target, name) {
```
Comment (EisenbergEffect) [inline src/resolvers.js]: Can we remove the `name` parameter here and then remove it from the places it's called below?

**Comment by doktordirk** [comment]
i wanted to sort the messy tests in different files, but oddly getting conflicts then. i'll make a separate PR for that later.

**Comment by doktordirk** [comment]
as for that last commit:
@inject as parameter decorator had the intermediate bug of not working with inheritance. the new implementation fixes that and also enables it for es6 (not that it is needed there) and without parameter and inferring the type from the metadata. So `@inject() public logger(Logger)` also works as i as user would expect .

in the lower part it allows inject as class member function decorator. i've no idea what use is that for and unfortunately the original commit provided neither explanation nor a test. the skeleton certainly works without that feature.

**Comment by EisenbergEffect** [comment]
@doktordirk This looks pretty good. Much simpler :) Anything else you want to do? If not, I'm ready to merge on our end. Let me know.

**Comment by bigopon** [comment]
This is nice, it also fixes a big related to autoinject fetching information from super class. Thats a bug we didnt notice until now.

**Comment by doktordirk** [comment]
@EisenbergEffect nothing important. some things, that could be done are:

* split up tests
* remove/reduce the typescript build errors which are not visible when importing
* remove unused and unusable factory.as
* remove (imho) useless @inject as class member function decorator

but maybe another PR would be better. i'll just move the most important StrategyResolver in resolvers up again (it got somewhat hidden at some stage), and i'm done here

**Comment by zewa666** [comment]
Great work @doktordirk

**Comment by Sayan751** [comment]
Is there any tentative plan on when to release this fix? I am having some [issue with unit-testing components associated with i18n](https://stackoverflow.com/questions/50586218/how-to-inject-element-to-unit-test-aurelia-view). I tried these changes on my development machine, and it seems to be working 🎉 Therefore I am particularly interested about the next release 😄

**Pull request merged**